### PR TITLE
Update login test in sync with other test

### PR DESCRIPTION
**Pull request by amitkrout**
## Was the change discussed in an issue?
fixes - part of #1473

## How to test changes?

make test-odo-login-e2e

### CODE REVIEW

File: scripts/openshiftci-presubmit-e2e-scenarios.sh [hunk 1]
```diff
@@ -14,4 +14,4 @@ export CUSTOM_HOMEDIR="/tmp/artifacts"
 
 make test-e2e-scenarios
 # Uncomment logout command once https://github.com/openshift/odo/issues/1836 is fixed
-odo logout
+# odo logout
```
Comment (cdrage) [inline scripts/openshiftci-presubmit-e2e-scenarios.sh]: https://github.com/openshift/odo/issues/1836 isn't fixed yet?

Comment (amitkrout) [reply scripts/openshiftci-presubmit-e2e-scenarios.sh]: Fixed via #1883

File: Makefile [hunk 2]
```diff
@@ -107,87 +114,121 @@ test:
 # Run generic integration tests
 .PHONY: test-generic
 test-generic:
+	odo login -u $(CLUSTER_USER_NAME) -p $(CLUSTER_PASSWORD) $(CLUSTER_LOGIN_URL) --insecure-skip-tls-verify
```
Comment (amitkrout) [inline Makefile]: FYI...
Due to open issue https://github.com/onsi/ginkgo/issues/440, i cannot make the ```odo login ...``` in suite level where login executed once globally before the spec.

Comment (kadel) [reply Makefile]: I think that makefile wouldn't ever contain `odo login`, or matter of fact, any other `odo` commands. 

Makefile should be environment agnostic.

**Comment by amitkrout** [comment]
> Agreed with @cdrage. This a not a good approach.

@kadel @cdrage The approach I was following like I followed in minishift. There the credential are passed through env variable and managed by Make file and then processed by the concern go file. Here in my changes I could not implemented env variable processing stuff through go file due to open issue https://github.com/onsi/ginkgo/issues/440 as i mentioned in one of my comment. My intention was to implement such a process which was very successful in minishift repo.

> How about separating the login and logout tests completely and leaving rest on the user.
What I mean by that is that only login and logout integration test will perform login and logout commands. Normal tests will expect that they are already logged in.

Yup, Works for me

> If the developer is running the test from terminal he/she will have to make sure to log in.
Scripts executing tests on CI will do the same.

+1, so i am closing the issue #1396 an it make no sense now onwards

> I think that makefile wouldn't ever contain odo login, or matter of fact, any other odo commands.

Absolutely, due to certain ginkgo constraint https://github.com/onsi/ginkgo/issues/440#issuecomment-511294531 i added the login stuff in Makefile thinking as a temporary change.

**Comment by kadel** [comment]
> Absolutely, due to certain ginkgo constraint [onsi/ginkgo#440 (comment)](https://github.com/onsi/ginkgo/issues/440#issuecomment-511294531) i added the login stuff in Makefile thinking as a temporary change.

login stuff shouldn't be in anywhere in the test suites (expect login integration test).
So the ginkgo constrains shouldn't affect anything.




> > If the developer is running the test from terminal he/she will have to make sure to log in.
> > Scripts executing tests on CI will do the same.
> 
> +1, so i am closing the issue #1396 an it make no sense now onwards

If everything else works, and I can just do 
```
oc login
make test-integration
```
sure lets close it

**Comment by amitkrout** [comment]
> > Absolutely, due to certain ginkgo constraint [onsi/ginkgo#440 (comment)](https://github.com/onsi/ginkgo/issues/440#issuecomment-511294531) i added the login stuff in Makefile thinking as a temporary change.
> 
> login stuff shouldn't be in anywhere in the test suites (expect login integration test).
> So the ginkgo constrains shouldn't affect anything.
> 
> > > If the developer is running the test from terminal he/she will have to make sure to log in.
> > > Scripts executing tests on CI will do the same.
> > 
> > 
> > +1, so i am closing the issue #1396 an it make no sense now onwards
> 
> If everything else works, and I can just do
> 
> ```
> oc login
> make test-integration
> ```
> 
> sure lets close it

Awesome!!!
Question: Why ```oc login``` why not ```odo login``` ?. I believe odo login works the same way as oc login does for normal user.

**Comment by kadel** [comment]
> Question: Why `oc login` why not `odo login` ?. I believe odo login works the same way as oc login does for normal user.

Yes, it should be the same. I just use `oc` because sometimes I don't have `odo` compiled yet :-)
Also, I usually use copy login commands from openshift web console. So it is just copy/paste for me,  and right now web console shows just the `oc` commands.
So no real reasons, just muscle memory and laziness :D

**Comment by cdrage** [comment]
> > > Absolutely, due to certain ginkgo constraint [onsi/ginkgo#440 (comment)](https://github.com/onsi/ginkgo/issues/440#issuecomment-511294531) i added the login stuff in Makefile thinking as a temporary change.
> > 
> > 
> > login stuff shouldn't be in anywhere in the test suites (expect login integration test).
> > So the ginkgo constrains shouldn't affect anything.
> > > > If the developer is running the test from terminal he/she will have to make sure to log in.
> > > > Scripts executing tests on CI will do the same.
> > > 
> > > 
> > > +1, so i am closing the issue #1396 an it make no sense now onwards
> > 
> > 
> > If everything else works, and I can just do
> > ```
> > oc login
> > make test-integration
> > ```
> > 
> > 
> > sure lets close it
> 
> Awesome!!!
> Question: Why `oc login` why not `odo login` ?. I believe odo login works the same way as oc login does for normal user.

I don't see anything with regards to using `oc login` and `oc logout` in Minishift: https://github.com/minishift/minishift/blob/master/Makefile

The problem with the current approach in this PR is that it assumes that the user is running `minishift` which may not be the case..

I do agree with @kadel that putting the `login` and `logout` commands in a separate test that can be ran by the user :+1:

**Comment by amitkrout** [comment]
> I don't see anything with regards to using oc login and oc logout in Minishift: https://github.com/minishift/minishift/blob/master/Makefile

This is not the specific use of ```oc login``` or ```oc logout``` this is more about how environment variable is being used smartly in the Makefile https://github.com/minishift/minishift/blob/master/Makefile#L264-L284.  As i mentioned above the change i have done i mean login and logout in Makefile was temporary due to issue https://github.com/onsi/ginkgo/issues/440#issuecomment-511294531 (https://github.com/openshift/odo/pull/1875#discussion_r303310349). Anyway we are using a separate testfile for login , however few usecase are missing (For example: odo logout) in the test file. For now i will update the test file with some more cases, later will fix with all possible usecase if any as part of #1845

**Comment by amitkrout** [comment]
@kadel @cdrage Done with the requested changes. PTAL

**Comment by amitkrout** [comment]
Ping @mohammedzee1000 - https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/1875/pull-ci-openshift-odo-master-unit/1441

This is new

File: docs/development.adoc [hunk 3]
```diff
@@ -324,7 +324,16 @@ $ TEST_EXEC_NODES=1 make test-source-e2e
...
 
-* For the entire integration test suite use:
+* For the login and logout command integration test:
++
+----
+$ make test-odo-login-e2e
+----
++
+
+NOTE: `make test-odo-login-e2e` doesn't honour enviornment variable `TEST_EXEC_NODES`. So by default it runs login and logout command integration test suite on a single ginkgo test node sequentially to avoid race condition in parallel run.
```
Comment (cdrage) [inline docs/development.adoc]: `environment`

File: docs/development.adoc [hunk 4]
```diff
@@ -324,7 +324,16 @@ $ TEST_EXEC_NODES=1 make test-source-e2e
 ----
 +
 
-* For the entire integration test suite use:
+* For the login and logout command integration test:
```
Comment (cdrage) [inline docs/development.adoc]: tests

File: docs/development.adoc (hunk 3 above)
Comment (cdrage) [reply docs/development.adoc]: `to avoid race conditions in a parallel run`

Comment (amitkrout) [reply docs/development.adoc]: Done

File: docs/development.adoc (hunk 4 above)
Comment (amitkrout) [reply docs/development.adoc]: Done

**Comment by cdrage** [comment]
Tests are failing.

Let's see if it's a false positive

/retest

**Comment by amitkrout** [comment]
ping @cdrage @kadel

**Comment by girishramnani** [comment]
I am just worried about the 3 runs that took us to get the CI to pass @amitkrout. might wanna reduce the concurrency?
cc @kadel 
removing the hold in the meantime
/hold cancel

**Comment by cdrage** [comment]
@girishramnani I agree with Girish with regards to the retesting. Is there a reason for the flakes?

**Comment by amitkrout** [comment]
> @girishramnani I agree with Girish with regards to the retesting. Is there a reason for the flakes?

@cdrage Yes, we have flake. These flake are from from OpenShift CI 
```
level=fatal msg="failed to fetch Terraform Variables: failed to fetch dependency of \"Terraform Variables\": failed to fetch dependency of \"Bootstrap Ignition Config\": failed to fetch dependency of \"Common Manifests\": failed to generate asset \"DNS Config\": getting public zone for \"origin-ci-int-aws.dev.rhcloud.com\": No public route53 zone found matching name \"origin-ci-int-aws.dev.rhcloud.com\""
```
and similar such failure which we don't have any control over it except retesting. Other flake is coming from an open issue #1832. However a recent flake (occurrence is very less) is introduced.
```
[odo]  ✗  Waiting for build to finish [30s]
[odo]  ✗  Failed to create component with name ref-test. Please use `odo config view` to view settings used to create component. Error: unable to wait for build ref-test-app-1 to run: timed out waiting for the condition 
```
Not sure is it a real threat or something need to be handled in odo code.These flakes are no where related to this pr change. I would say please review. 

Anyway we are talking these flake in separate issues cc_ @kadel

**Comment by amitkrout** [comment]
> I am just worried about the 3 runs that took us to get the CI to pass @amitkrout. might wanna reduce the concurrency?

@girishramnani please refer https://github.com/openshift/odo/pull/1909#issuecomment-511667494 and https://github.com/openshift/odo/pull/1909#issuecomment-511729801 and feel fee to add your opinion.

**Comment by amitkrout** [comment]
/test benchmark
ping @kadel @cdrage

File: docs/development.adoc [hunk 5]
```diff
@@ -187,165 +187,174 @@ Refer to the odo clean test link:https://github.com/openshift/odo/blob/master/te
 
 Integration tests can be run in the following two ways:
 
-* To run the test in parallel (default: 4 ginkgo test node), on a test cluster:
+* To run the tests in parallel (default: 4 ginkgo test node), on a test cluster:
```
Comment (mik-dass) [inline docs/development.adoc]: `default: 2`

**Comment by amitkrout** [comment]
> > I am just worried about the 3 runs that took us to get the CI to pass @amitkrout. might wanna reduce the concurrency?
> 
> @girishramnani please refer [#1909 (comment)](https://github.com/openshift/odo/pull/1909#issuecomment-511667494) and [#1909 (comment)](https://github.com/openshift/odo/pull/1909#issuecomment-511729801) and feel free to add your opinion.

Decrease the test node to 2 as per our cabal discussion

**Pull request merged**